### PR TITLE
feat: oracle manager module

### PR DESCRIPTION
**Pull request by Rickk137**

### CODE REVIEW

File: packages/oracle-manager/contracts/interfaces/IOracleManagerModule.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    function getNodeId(
+        bytes32[] memory parents,
+        NodeDefinition.NodeType nodeType,
+        bytes memory parameters
+    ) external returns (bytes32);
+
+    function getNode(bytes32 nodeId) external view returns (NodeDefinition.Data memory);
+
+    function process(bytes32 nodeId) external view returns (Node.Data memory);
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/interfaces/IOracleManagerModule.sol:23]: adding /// @notice with some description will help build out docs

File: packages/oracle-manager/contracts/mocks/AggregatorV3Mock.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../interfaces/external/IAggregatorV3Interface.sol";
+
+contract AggregatorV3Mock is IAggregatorV3Interface {
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/mocks/AggregatorV3Mock.sol:6]: can you just import the mock from `packages/synthetix-main` and inherit here?

File: packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import "../../utils/ExternalNodeLibrary.sol";
+import "../../utils/PythNodeLibrary.sol";
+import "../../utils/ChainlinkNodeLibrary.sol";
+
+import "../../storage/Node.sol";
+import "../../storage/NodeDefinition.sol";
+
+contract OracleManagerModule is IOracleManagerModule {
+    error UnsupportedNodeType(uint nodeType);
+    error NodeNotRegistered(bytes32 nodeId);
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol:15]: move these to the interface?  seems like that's the pattern i've noticed.  nit pick :)

File: packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        for (uint256 i = 0; i < nodeDefinition.parents.length; i++) {
+            if (!_nodeIsRegistered(nodeDefinition.parents[i])) {
+                revert NodeNotRegistered(nodeDefinition.parents[i]);
+            }
+        }
+
+        (, nodeId) = NodeDefinition.create(nodeDefinition);
+    }
+
+    function _nodeIsRegistered(bytes32 nodeId) internal view returns (bool) {
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol:95]: rename to `_isNodeRegistered`

File: packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    modifier onlyValidNodeType(NodeDefinition.NodeType nodeType) {
+        if (!_validateNodeType(nodeType)) {
+            revert UnsupportedNodeType(uint(nodeType));
+        }
+
+        _;
+    }
+
+    function _validateNodeType(NodeDefinition.NodeType nodeType) internal pure returns (bool) {
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol:65]: can just move this functionality into the modifier; no reason for a layer of misdirection

File: packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        if (
+            NodeDefinition.NodeType.REDUCER == nodeType ||
+            NodeDefinition.NodeType.EXTERNAL == nodeType ||
+            NodeDefinition.NodeType.CHAINLINK == nodeType ||
+            NodeDefinition.NodeType.PYTH == nodeType
+        ) return true;
+
+        return false;
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol:73]:
```suggestion
       return 
            NodeDefinition.NodeType.REDUCER == nodeType ||
            NodeDefinition.NodeType.EXTERNAL == nodeType ||
            NodeDefinition.NodeType.CHAINLINK == nodeType ||
            NodeDefinition.NodeType.PYTH == nodeType
       
```

File: packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    function _getNodeId(NodeDefinition.Data memory nodeDefinition) internal pure returns (bytes32) {
+        return NodeDefinition.getId(nodeDefinition);
+    }
+
+    function _registerNode(NodeDefinition.Data memory nodeDefinition)
+        internal
+        onlyValidNodeType(nodeDefinition.nodeType)
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol:82]: i'd suggest moving this modifier to `registerNode` and passing in the `NodeType` that's passed in.

File: packages/oracle-manager/contracts/utils/ReducerNodeLibrary.sol [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    function min(Node.Data[] memory prices) internal pure returns (Node.Data memory minPrice) {
+        minPrice = prices[0];
+        for (uint256 i = 0; i < prices.length; i++) {
+            if (prices[i].price < minPrice.price) {
+                minPrice = prices[i];
+            }
+        }
+    }
+
+    function quickSort(
```
Comment (sunnyvempati) [inline packages/oracle-manager/contracts/utils/ReducerNodeLibrary.sol:79]: hah. fun.

**Review by sunnyvempati — COMMENTED** [review]
awesome stuff; threw in some suggestions 👍🏽

**Comment by sunnyvempati** [comment]
another note @Rickk137 

let's add events to all external calls.  thank you!

File: packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol (hunk 3 above)
Comment (dbeal-eth) [reply packages/oracle-manager/contracts/modules/core/OracleManagerModule.sol:15]: i dont think we have been moving the errors to the interface, I'm not sure why though.

File: packages/oracle-manager/test/integration/modules/OracleManagerModule/OracleManagerModule.nodes.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import assert from 'assert/strict';
+import assertBn from '@synthetixio/core-utils/utils/assertions/assert-bignumber';
+import { ethers } from 'ethers';
+
+import { bootstrapWithNodes } from '../../bootstrap';
+import NodeTypes from '../../mixins/Node.types';
+import NodeOperations from '../../mixins/Node.operations';
+import assertRevert from '@synthetixio/core-utils/utils/assertions/assert-revert';
+
+describe('Oracle manager module nodes', function () {
```
Comment (dbeal-eth) [inline packages/oracle-manager/test/integration/modules/OracleManagerModule/OracleManagerModule.nodes.ts:10]: I should have come in earlier to comment on this ,but the format of these tests does not correspond to the v3 testing strategy cc @ajsantander . We will probably want to update these in the near future.

**Review by dbeal-eth — APPROVED** [review]

**Comment by Rickk137** [comment]
> another note @Rickk137
> 
> let's add events to all external calls. thank you!

Good point. The only place that i can think of raising an event is registerNode. Do you think we need an event on other actions?

**Comment by Rickk137** [comment]
> @Rickk137 is this ready for merge?

Yes i've resolved the comments

**Pull request merged**